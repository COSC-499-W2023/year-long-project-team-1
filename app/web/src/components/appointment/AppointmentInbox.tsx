export const AppointmentInbox = () => {
  return (
    <div>
      <h1>Appointment Inbox</h1>
    </div>
  );
};
