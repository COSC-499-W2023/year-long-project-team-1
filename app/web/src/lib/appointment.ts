import { User } from "@prisma/client";

export interface ViewableAppointment {
  id: number;
  clientUser: Partial<User> | null;
  professionalUser: Partial<User> | null;
  time: Date;
}
