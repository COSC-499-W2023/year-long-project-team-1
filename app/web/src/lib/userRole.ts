export enum UserRole {
  PROFESSIONAL = "professional",
  CLIENT = "client",
}
