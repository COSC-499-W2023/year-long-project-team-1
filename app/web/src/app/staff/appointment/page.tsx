/*
 * Copyright [2023] [Privacypal Authors]
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  findUserById,
  findUserSanitizedById,
  getLoggedInUser,
  getProfessionalAppointment,
} from "@app/actions";
import AppointmentViewer from "@components/appointment/AppointmentViewer";
import { ViewableAppointment } from "@lib/appointment";
import { NextPageProps } from "@lib/url";
import { notFound, redirect } from "next/navigation";

export default async function AppointmentPage({ searchParams }: NextPageProps) {
  // if no appointment id, redirect to staff dashboard
  if (!searchParams?.id || Array.isArray(searchParams?.id)) {
    return redirect("/staff/");
  }

  try {
    // parseInt will throw an error if the id is not a number-string
    const apptId = parseInt(searchParams?.id);
    const user = await getLoggedInUser();

    // if somehow there is no user, redirect to login
    if (!user) redirect("/login");

    // this function will throw an error if the logged in user is not a professional
    const appt = await getProfessionalAppointment(user, apptId);

    // if there is no appointment by that id, return 404
    if (!appt) return notFound();

    // display appointment data
    const client = await findUserSanitizedById(appt.clientId);

    if (!client) notFound();

    const viewableAppointment: ViewableAppointment = {
      id: appt.id,
      clientUser: client,
      professionalUser: user,
      time: appt.time,
    };

    return (
      <AppointmentViewer appointment={viewableAppointment} viewer={user} />
    );
  } catch (error: any) {
    // return 404 if there is an error
    return notFound();
  }
}
