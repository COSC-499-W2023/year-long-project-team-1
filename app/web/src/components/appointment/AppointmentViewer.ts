"use client";

import LinkButton from "@components/form/LinkButton";
import { ViewableAppointment } from "@lib/appointment";
import { Button, Card, CardBody } from "@patternfly/react-core";
import { Role, User } from "@prisma/client";

interface AppointmentViewerProps {
  appointment: ViewableAppointment;
  viewer?: Omit<User, "password">;
}

export default function AppointmentViewer({
  appointment,
  viewer,
}: AppointmentViewerProps) {
  const proUser =
    appointment.professionalUser.firstname +
    " " +
    appointment.professionalUser.lastname;
  const clientUser =
    appointment.clientUser.firstname + " " + appointment.clientUser.lastname;

  const viewerRole = viewer?.role;

  const uploadButton =
    viewerRole === Role.CLIENT ? (
      <LinkButton href="/upload" label="Upload a video to this appointment" />
    ) : null;

  return (
    <Card>
      <CardBody>
        <h2>Appointment</h2>
        <time>{appointment.time.toLocaleDateString()}</time>
        <p>Professional: {proUser}</p>
        <p>Client: {clientUser}</p>
        {uploadButton}
      </CardBody>
    </Card>
  );
}
