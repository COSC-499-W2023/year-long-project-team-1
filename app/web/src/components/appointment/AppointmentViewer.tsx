/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use client";

import LinkButton from "@components/form/LinkButton";
import { ViewableAppointment } from "@lib/appointment";
import { Button, Card, CardBody } from "@patternfly/react-core";
import { Role, User } from "@prisma/client";
import { Session } from "next-auth";

interface AppointmentViewerProps {
  appointment: ViewableAppointment;
  viewer?: Omit<Session["user"], "password">;
}

export default function AppointmentViewer({
  appointment,
  viewer,
}: AppointmentViewerProps) {
  const proUser =
    appointment.professionalUser?.firstName +
    " " +
    appointment.professionalUser?.lastName;
  const clientUser =
    appointment.clientUser?.firstName + " " + appointment.clientUser?.lastName;

  const viewerRole = viewer?.role;

  const uploadButton =
    viewerRole === Role.CLIENT ? (
      <LinkButton
        href={`/upload?id=${appointment.id}`}
        label="Upload a video to this appointment"
      />
    ) : null;

  return (
    <Card>
      <CardBody>
        <h2>Appointment</h2>
        <time>{appointment.time.toLocaleDateString()}</time>
        <p>Professional: {proUser}</p>
        <p>Client: {clientUser}</p>
        {uploadButton}
      </CardBody>
    </Card>
  );
}
