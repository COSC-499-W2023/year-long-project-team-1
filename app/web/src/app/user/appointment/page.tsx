import {
  findUserById,
  findUserSanitizedById,
  getLoggedInUser,
  getClientAppointment,
} from "@app/actions";
import AppointmentViewer from "@components/appointment/AppointmentViewer";
import { ViewableAppointment } from "@lib/appointment";
import { NextPageProps } from "@lib/url";
import { notFound, redirect } from "next/navigation";

export default async function AppointmentPage({ searchParams }: NextPageProps) {
  // if no appointment id, redirect to staff dashboard
  if (!searchParams?.id || Array.isArray(searchParams?.id)) {
    console.log("no id");
    return redirect("/staff/");
  }

  try {
    // parseInt will throw an error if the id is not a number-string
    const apptId = parseInt(searchParams?.id);
    const user = await getLoggedInUser();

    // if somehow there is no user, redirect to login
    if (!user) redirect("/login");

    // this function will throw an error if the logged in user is not a professional
    const appt = await getClientAppointment(user, apptId);

    // if there is no appointment by that id, return 404
    if (!appt) return notFound();

    // display appointment data
    const professional = await findUserSanitizedById(appt.proId);

    if (!professional) notFound();

    const viewableAppointment: ViewableAppointment = {
      id: appt.id,
      clientUser: user,
      professionalUser: professional,
      time: appt.time,
    };

    return (
      <AppointmentViewer appointment={viewableAppointment} viewer={user} />
    );
  } catch (error: any) {
    // return 404 if there is an error
    console.error(error);
    return notFound();
  }
}
