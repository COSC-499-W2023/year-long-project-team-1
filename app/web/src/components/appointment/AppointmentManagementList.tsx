"use client";

import { User } from "@prisma/client";
import { useEffect, useState } from "react";
import { getAllProfessionalAppointmentDetails } from "@app/actions";
import AppointmentViewer from "./AppointmentViewer";

interface AppointmentManagementListProps {
  professional: User;
}

export default function AppointmentManagementList({
  professional,
}: AppointmentManagementListProps) {
  const [appointments, setAppointments] = useState<JSX.Element[]>([]);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    getAllProfessionalAppointmentDetails(professional).then((i) => {
      let componentList: JSX.Element[] = [];
      i.forEach((j, count) => {
        componentList.push(
          <AppointmentViewer
            appointment={j}
            viewer={professional}
            key={count}
          />,
        );
      });
      setAppointments(componentList);
      setLoading(false);
    });
  }, []);

  return (
    <main>
      {loading ? (
        <div>Loading...</div>
      ) : appointments.length > 0 ? (
        appointments
      ) : (
        "No appointments found."
      )}
    </main>
  );
}
