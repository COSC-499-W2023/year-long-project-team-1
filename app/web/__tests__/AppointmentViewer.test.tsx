/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import AppointmentViewer from "@components/appointment/AppointmentViewer";
import { render, screen } from "@testing-library/react";
import { ViewableAppointment } from "@lib/appointment";
import { UserRole } from "@lib/userRole";

describe("AppointmentViewer Component", () => {
  const pro = {
    id: 1,
    username: "professional_test_user",
    email: "pro@example.com",
    firstName: "Pro",
    lastName: "User",
    role: UserRole.PROFESSIONAL,
  };
  const client = {
    id: 2,
    username: "client_test_user",
    email: "client@example.com",
    firstName: "Client",
    lastName: "User",
    role: UserRole.CLIENT,
  };
  const appt: ViewableAppointment = {
    id: 1,
    professionalUser: pro,
    clientUser: client,
    time: new Date(Date.UTC(2025, 0, 1, 12, 0, 0)), // jan 1st 2025 12:00 noon
    video_count: 1,
  };

  it("component is unchanged from snapshot", () => {
    const { container: container } = render(
      <div>
        <AppointmentViewer appointment={appt} viewer={pro} />,
        <AppointmentViewer appointment={appt} viewer={client} />
      </div>,
    );
    expect(container).toMatchSnapshot();
  });
  it("pro viewer has cancel and view details buttons", () => {
    const { container: proContainer } = render(
      <AppointmentViewer appointment={appt} viewer={pro} />,
    );
    expect(screen.getByText("Cancel appointment")).toBeDefined();
    expect(screen.getByText("View appointment details")).toBeDefined();
  });
  it("client viewer has upload video button", () => {
    const { container: clientContainer } = render(
      <AppointmentViewer appointment={appt} viewer={client} />,
    );
    expect(
      screen.getByText("Upload a video to this appointment"),
    ).toBeDefined();
  });
});
