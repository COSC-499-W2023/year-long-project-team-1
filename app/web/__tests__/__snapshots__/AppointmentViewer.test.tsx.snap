// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentViewer Component component is unchanged from snapshot 1`] = `
<div>
  <div>
    <div
      class="pf-v5-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-v5-c-card__body"
      >
        <h2>
          Appointment
        </h2>
        <time>
          2025-01-01
        </time>
        <p>
          Professional: 
          Pro User
        </p>
        <p>
          Client: 
          Client User
        </p>
        <p>
          Number of associated videos: 
          1
        </p>
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-danger"
          data-ouia-component-id="OUIA-Generated-Button-danger-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Cancel appointment
        </button>
        <a
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary"
          data-ouia-component-id="OUIA-Generated-Button-primary-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          href="/staff/appointment?id=1"
        >
          View appointment details
        </a>
      </div>
    </div>
    ,
    <div
      class="pf-v5-c-card"
      data-ouia-component-id="OUIA-Generated-Card-2"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-v5-c-card__body"
      >
        <h2>
          Appointment
        </h2>
        <time>
          2025-01-01
        </time>
        <p>
          Professional: 
          Pro User
        </p>
        <p>
          Client: 
          Client User
        </p>
        <p>
          Number of associated videos: 
          1
        </p>
        <a
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary"
          data-ouia-component-id="OUIA-Generated-Button-primary-2"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          href="/upload"
        >
          Upload a video to this appointment
        </a>
      </div>
    </div>
  </div>
</div>
`;
